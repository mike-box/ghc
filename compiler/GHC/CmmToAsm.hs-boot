module GHC.CmmToAsm
where

import GHC.Prelude

import GHC.IO.Handle.Types
import GHC.CmmToAsm.Config
import GHC.Types.Unique.Supply
import GHC.Utils.Logger
import GHC.Unit.Module.Location
import GHC.Data.Stream
import GHC.Cmm

nativeCodeGen :: forall a . Logger -> NCGConfig -> ModLocation -> Handle -> UniqSupply
              -> Stream IO RawCmmGroup a
              -> IO a
