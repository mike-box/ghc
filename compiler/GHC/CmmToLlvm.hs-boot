module GHC.CmmToLlvm
where

import GHC.Prelude

import GHC.IO.Handle.Types
import {-# SOURCE #-} GHC.CmmToLlvm.Config (LlvmCgConfig)
import GHC.Utils.Logger
import GHC.Data.Stream
import GHC.Cmm

llvmCodeGen :: Logger -> LlvmCgConfig -> Handle
               -> Stream IO RawCmmGroup a
               -> IO a
