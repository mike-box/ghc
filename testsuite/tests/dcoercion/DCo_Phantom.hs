{-# LANGUAGE TypeInType, TypeOperators, TypeFamilies,
             UndecidableInstances, ConstraintKinds #-}
module DCo_Phantom where

import GHC.TypeLits as L
import Data.Type.Bool


-- We define a very simplistic O notation, with sufficient expressiveness
-- to capture the complexity of a few simple sorting algorithms
data AsympPoly = NLogN Nat Nat

-- Synonyms for common terms
type N     = NLogN 1 0
type LogN  = NLogN 0 1
type One   = NLogN 0 0

-- Just to be able to write it nicely
type O (a :: AsympPoly) = a

type family (^.) (n :: AsympPoly) (m :: Nat) :: AsympPoly where
  (NLogN a b) ^. n = (NLogN (a L.* n) (b L.* n))

infixl 7 ^.

newtype Sorted (cpu :: AsympPoly) -- The minimum operational complexity
                                       -- this algorithm satisfies.
               (mem :: AsympPoly) -- The minimum space complexity this
                                       -- algorithm satisfies.
               (stable :: Bool)        -- Whether the sort is stable or not.
               a                       -- What was being sorted.
               = Sorted {sortedBy :: [a]}

mySortA :: Sorted (O(N^.2)) (O(N)) True Integer
mySortA = _a [3,1,2]
