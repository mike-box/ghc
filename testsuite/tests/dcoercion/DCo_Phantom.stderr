
DCo_Phantom.hs:35:11: warning: [-Wtyped-holes (in -Wdefault)]
    • Found hole:
        _a :: [Integer] -> Sorted (O ('NLogN 2 0)) (O N) 'True Integer
      Or perhaps ‘_a’ is mis-spelled, or not in scope
    • In the expression: _a [3, 1, 2]
      In an equation for ‘mySortA’: mySortA = _a [3, 1, 2]
    • Relevant bindings include
        mySortA :: Sorted (O (N ^. 2)) (O N) 'True Integer
          (bound at DCo_Phantom.hs:35:1)
      Valid hole fits include
        Sorted :: forall (cpu :: AsympPoly) (mem :: AsympPoly)
                         (stable :: Bool) a.
                  [a] -> Sorted cpu mem stable a
          with Sorted @(O ('NLogN 2 0)) @(O N) @'True @Integer
          (defined at DCo_Phantom.hs:32:18)
